server/src/main/java/org/opensearch/search/aggregations/bucket/histogram/DateHistogramAggregator.java

```diff
@@ -126,7 +126,8 @@ class DateHistogramAggregator extends BucketsAggregator implements SizedBucketAg
             )
         );
         if (fastFilterContext.isRewriteable(parent, subAggregators.length)) {
-            fastFilterContext.buildFastFilter();
+            fastFilterContext.setFieldName(valuesSourceConfig.fieldType().name());
+            fastFilterContext.buildRanges();
```

Comment:
At shard level, we check the provided query, whether it's a match all or range query, then use it to build ranges.
https://github.com/opensearch-project/OpenSearch/blob/1765e20cf3939cb0fff0b8af37294da26d9cbea0/server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java#L135-L152

At segment level, we only proceed if ranges are not built at shard level and segment matches all.
https://github.com/opensearch-project/OpenSearch/blob/1765e20cf3939cb0fff0b8af37294da26d9cbea0/server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java#L444

> avoid building ranges for segments where this optimization cannot be applied

I'm not worried about the performance impact of this. We can also think build at shard level is better since ranges built at segment level will always be subset of the ranges build at shard level.